zcash_primitives/src/transaction/txid.rs

```diff
@@ -0,0 +1,300 @@
+            ch.write_all(&s_out.enc_ciphertext[..52]).unwrap();
+
+            mh.write_all(&s_out.enc_ciphertext[52..564]).unwrap();
+
+            nh.write_all(&s_out.cv.to_bytes()).unwrap();
+            nh.write_all(&s_out.enc_ciphertext[564..]).unwrap();
+            nh.write_all(&s_out.out_ciphertext).unwrap();
+        }
+
+        h.write_all(ch.finalize().as_bytes()).unwrap();
+        h.write_all(mh.finalize().as_bytes()).unwrap();
+        h.write_all(nh.finalize().as_bytes()).unwrap();
+    }
+    h.finalize()
+}
+
+/// Implements [ZIP 244 section T.1](https://zips.z.cash/zip-0244#t-1-header-digest)
+fn hash_header_txid_data(
+    version: TxVersion,
+    // we commit to the consensus branch ID with the header
+    consensus_branch_id: BranchId,
+    lock_time: u32,
+    expiry_height: BlockHeight,
+) -> Blake2bHash {
+    let mut h = hasher(ZCASH_HEADERS_HASH_PERSONALIZATION);
+
+    h.write_u32::<LittleEndian>(version.header()).unwrap();
+    h.write_u32::<LittleEndian>(version.version_group_id())
+        .unwrap();
+    h.write_u32::<LittleEndian>(consensus_branch_id.into())
+        .unwrap();
+    h.write_u32::<LittleEndian>(lock_time).unwrap();
+    h.write_u32::<LittleEndian>(expiry_height.into()).unwrap();
+
+    h.finalize()
+}
+
+/// Implements [ZIP 244 section T.2](https://zips.z.cash/zip-0244#t-2-transparent-digest)
+pub(crate) fn hash_transparent_txid_data(txdata: &TransactionData) -> Blake2bHash {
+    let mut h = hasher(ZCASH_TRANSPARENT_HASH_PERSONALIZATION);
+    if !txdata.vin.is_empty() || !txdata.vout.is_empty() {
+        h.write_all(transparent_prevout_hash(&txdata.vin).as_bytes())
+            .unwrap();
+        h.write_all(transparent_sequence_hash(&txdata.vin).as_bytes())
+            .unwrap();
+        h.write_all(transparent_outputs_hash(&txdata.vout).as_bytes())
+            .unwrap();
+    }
+    h.finalize()
+}
+
+/// Implements [ZIP 244 section T.3](https://zips.z.cash/zip-0244#t-3-sapling-digest)
+fn hash_sapling_txid_data(txdata: &TransactionData) -> Blake2bHash {
+    let mut h = hasher(ZCASH_SAPLING_HASH_PERSONALIZATION);
+    if !(txdata.shielded_spends.is_empty() && txdata.shielded_outputs.is_empty()) {
+        h.write_all(hash_sapling_spends(&txdata.shielded_spends).as_bytes())
+            .unwrap();
+
+        h.write_all(hash_sapling_outputs(&txdata.shielded_outputs).as_bytes())
+            .unwrap();
+
+        h.write_all(&txdata.value_balance.to_i64_le_bytes())
+            .unwrap();
+    }
+    h.finalize()
+}
+
+#[cfg(feature = "zfuture")]
+fn hash_tze_txid_data(txdata: &TransactionData) -> Blake2bHash {
+    let mut h = hasher(ZCASH_TZE_HASH_PERSONALIZATION);
+    if !(txdata.tze_inputs.is_empty() && txdata.tze_outputs.is_empty()) {
+        h.write_all(hash_tze_inputs(&txdata.tze_inputs).as_bytes())
+            .unwrap();
+        h.write_all(hash_tze_outputs(&txdata.tze_outputs).as_bytes())
+            .unwrap();
```

Comment:
I note that this is missing the [per-input digest](https://github.com/zcash/librustzcash/blob/63f22ade9f92ddc23023e17d2413a53b295a5a07/zcash_primitives/src/transaction/txid.rs#L279-L281), but I'm fine with that because the mobile SDKs won't have this code enabled.